File: daemon/src/services/cloudflared.ts
import { execFileSync, spawn } from "node:child_process";
import type { ChildProcess } from "node:child_process";
import * as tunnelStore from "../state/tunnel-store.js";

interface TunnelState {
  enabled: boolean;
  tunnelUrl: string | null;
  process: ChildProcess | null;
}

const state: TunnelState = { enabled: false, tunnelUrl: null, process: null };

/**
 * Spawn in progress. `state.enabled` stays false for up to 10 s while cloudflared
 * boots, so without this latch a second enable() in that window spawns a second
 * cloudflared, overwrites `state.process`, and orphans the first — leaving a live
 * public tunnel that `disable()` can never kill.
 */
let pending: Promise<{ tunnelUrl: string }> | null = null;

/**
 * The in-flight spawn attempt, if any. `disable()`/`shutdownKill()` must cancel
 * this (not just kill the process) — otherwise the stdio listeners keep
 * scanning output for the rest of the 10s window, `pending` never clears
 * (wedging the next `enable()` until the timeout fires), and a URL that happens
 * to appear in that window resolves with a pid that's already dead.
 */
let activeSpawn: { child: ChildProcess; cancel: (reason: string) => void } | null = null;

const TUNNEL_URL_RE = /https:\/\/[a-z0-9-]+\.trycloudflare\.com/;
const SPAWN_TIMEOUT_MS = 10_000;

// Deliberately distinct from the 500ms advanceTimersByTimeAsync idiom already
// used throughout cloudflared.test.ts — sharing that value would make the
// escalation timer fire mid-advance in any test whose mocked pgrep returns
// real pids, entangling unrelated test assertions with sweep internals.
const SWEEP_GRACE_MS = 2000;

/** Spawn cloudflared and return the public tunnel URL. */
export function enable(port: number): Promise<{ tunnelUrl: string }> {
  if (state.enabled && state.tunnelUrl) {
    return Promise.resolve({ tunnelUrl: state.tunnelUrl });
  }
  if (pending) return pending;

  pending = spawnTunnel(port);
  // Clear the latch either way so a failed attempt can be retried.
  pending.then(
    () => { pending = null; },
    () => { pending = null; },
  );
  return pending;
}

function spawnTunnel(port: number): Promise<{ tunnelUrl: string }> {
  // Sweep first, before anything else — only reached when enable() has
  // decided to actually spawn (never on its "already enabled" early return,
  // which must not kill the live tunnel it's about to report as healthy).
  sweepOrphans(port);
  return new Promise((resolve, reject) => {
    const child = spawn(process.env.VST_CLOUDFLARED_BIN ?? "cloudflared", ["tunnel", "--url", `http://127.0.0.1:${port}`], {
      stdio: ["ignore", "pipe", "pipe"],
    });
    state.process = child;

    let resolved = false;

    function finishResolve(url: string) {
      if (resolved) return;
      resolved = true;
      clearTimeout(timer);
      if (activeSpawn?.child === child) activeSpawn = null;
      state.enabled = true;
      state.tunnelUrl = url;
      tunnelStore.setState({
        enabled: true,
        currentUrl: url,
        currentPid: child.pid ?? null,
        startedAt: Date.now(),
        port,
      });
      resolve({ tunnelUrl: url });
    }

    function finishReject(err: Error) {
      if (resolved) return;
      resolved = true;
      clearTimeout(timer);
      if (activeSpawn?.child === child) activeSpawn = null;
      reject(err);
    }

    // Accumulate across chunks rather than testing each one in isolation:
    // stdout/stderr are byte streams with no message framing, so the URL can
    // (and on a slow pipe does) arrive split across two `data` events —
    // "https://abc-" then "def.trycloudflare.com" — which no per-chunk regex
    // would ever match, wedging the spawn until the 10s timeout. Capped so a
    // chatty cloudflared can't grow this unboundedly during the spawn window;
    // the tail is retained (not cleared) so a URL straddling the cap boundary
    // still matches.
    const MAX_BUFFER = 64 * 1024;
    let buffer = "";

    function onData(chunk: Buffer) {
      buffer += chunk.toString("utf8");
      if (buffer.length > MAX_BUFFER) buffer = buffer.slice(-MAX_BUFFER);
      const match = TUNNEL_URL_RE.exec(buffer);
      if (match) finishResolve(match[0]);
    }

    child.stdout?.on("data", onData);
    child.stderr?.on("data", onData);

    const timer = setTimeout(() => {
      child.kill();
      finishReject(new Error("cloudflared did not emit a URL within 10s"));
    }, SPAWN_TIMEOUT_MS);

    // Lets killTrackedProcess() (disable()/shutdownKill()) cancel THIS attempt
    // if it fires mid-spawn, instead of leaving it to wedge `pending` and poll
    // for the rest of SPAWN_TIMEOUT_MS after the process is already dead.
    activeSpawn = {
      child,
      cancel: (reason) => finishReject(new Error(reason)),
    };

    child.on("error", (err) => {
      // Guard by identity, not unconditionally — an error event for a process
      // that's already been superseded by a newer spawn must not clobber the
      // newer one's tracking (same hazard as the exit handler below).
      if (state.process === child) state.process = null;
      const msg = (err as NodeJS.ErrnoException).code === "ENOENT"
        ? "cloudflared not found — install it and ensure it is on PATH (run: vst doctor)"
        : err.message;
      finishReject(new Error(msg));
    });

    child.on("exit", (code) => {
      // Guard by identity, not pid: a late exit event for a process that has
      // already been superseded by a newer spawn (disable() + enable() in
      // quick succession) must never clobber the newer one's state.
      if (state.process !== child) return;
      state.process = null;

      if (!resolved) {
        finishReject(new Error(`cloudflared exited with code ${code ?? "?"} before emitting URL`));
      } else {
        // Process died after URL was emitted — tunnel is gone.
        state.enabled = false;
        state.tunnelUrl = null;
        const cur = tunnelStore.getState();
        if (cur.currentPid === child.pid) tunnelStore.clearProcess();
        console.warn("[cloudflared] tunnel process exited unexpectedly");
      }
    });
  });
}

/**
 * Enumerate every real OS process matching the exact cloudflared invocation
 * this daemon uses for `port`, including the port itself — orphan-reconciliation
 * sweep (see cloudflared-orphan-sweep plan, Decision 3). Matches the full argv,
 * not just the binary name, so an unrelated user-run cloudflared tunnel on a
 * different port/purpose is never touched. Dots are regex-escaped and the
 * pattern is `$`-anchored on the port so e.g. `:655` can't substring-match a
 * real tunnel on `:65535`. Fail-open on any `pgrep` error (missing binary,
 * sandboxed, etc.) — a sweep that finds nothing is safe, never crashing the
 * caller is the priority, same convention as `isLikelyCloudflaredProcess` below.
 */
function findMatchingPids(port: number): number[] {
  const pattern = `cloudflared tunnel --url http://127\\.0\\.0\\.1:${port}$`;
  try {
    const out = execFileSync("pgrep", ["-f", pattern], { encoding: "utf8", timeout: 2000 }).trim();
    if (!out) return [];
    return out
      .split("\n")
      .map((line) => Number.parseInt(line.trim(), 10))
      .filter((n) => Number.isFinite(n));
  } catch (err) {
    // pgrep's documented contract: exit 1 == "no processes matched" — not an error.
    if ((err as NodeJS.ErrnoException & { status?: number }).status === 1) return [];
    console.warn("[cloudflared] sweep: pgrep enumeration failed:", err);
    return [];
  }
}

/** ESRCH-based liveness probe — `process.kill(pid, 0)` signals nothing, just checks existence. */
function isProcessAlive(pid: number): boolean {
  try {
    process.kill(pid, 0);
    return true;
  } catch {
    return false;
  }
}

/**
 * OS-truth reconciliation sweep — treats the OS as the source of truth for
 * "what's running," not the DB (cloudflared-orphan-sweep plan, report
 * "Proposed" diagram). Runs at every entry point that changes tunnel state,
 * BEFORE anything else: enumerates every real cloudflared process bound to
 * `port`, SIGTERMs all of them unconditionally (no "is this mine" branch —
 * a fresh boot/enable() always mints a brand-new URL, so nothing found here
 * is ever worth keeping), then escalates any still-alive pid to SIGKILL after
 * a grace period. Additive to (does not replace) the existing single-pid
 * `tunnel_state`-derived cleanup below.
 *
 * The SIGKILL escalation is deliberately fire-and-forget: cloudflared dials
 * out to the local port rather than binding it, so a lingering old process
 * can't conflict with a freshly-spawned one, and blocking the new spawn on
 * old-process death buys nothing but latency. The escalation timer is
 * `.unref()`'d so it can never keep the daemon process alive on its own.
 */
function sweepOrphans(port: number): void {
  const pids = findMatchingPids(port);
  if (pids.length === 0) return;
  for (const pid of pids) {
    console.log(`[cloudflared] sweep: SIGTERM pid ${pid} (port ${port})`);
    try {
      process.kill(pid, "SIGTERM");
    } catch (err) {
      if ((err as NodeJS.ErrnoException).code !== "ESRCH") {
        console.warn(`[cloudflared] sweep: failed to SIGTERM pid ${pid}:`, err);
      }
    }
  }
  setTimeout(() => {
    for (const pid of pids) {
      if (!isProcessAlive(pid)) continue;
      // Re-check identity before SIGKILL too — same pid-reuse hazard
      // isLikelyCloudflaredProcess() already guards against for persisted
      // pids below; the grace period is long enough for a pid to have been
      // recycled by an unrelated process.
      if (!isLikelyCloudflaredProcess(pid)) continue;
      console.warn(`[cloudflared] sweep: pid ${pid} still alive after grace period — SIGKILL`);
      try {
        process.kill(pid, "SIGKILL");
      } catch (err) {
        if ((err as NodeJS.ErrnoException).code !== "ESRCH") {
          console.warn(`[cloudflared] sweep: failed to SIGKILL pid ${pid}:`, err);
        }
      }
    }
  }, SWEEP_GRACE_MS).unref();
}

/**
 * Best-effort identity check before signaling a pid we only know from a
 * persisted record (not our own live `ChildProcess` handle) — after a reboot
 * or a long gap, that pid can easily have been reused by an unrelated
 * process, and blindly SIGTERM-ing it would kill someone else's process.
 * `ps` (not `/proc`) is used since it works on both Linux and macOS.
 * Inconclusive (ps missing, sandboxed, pid already gone) → fail open and
 * attempt the kill anyway; `process.kill`'s own ESRCH handling is still the
 * backstop for "it's already gone".
 */
function isLikelyCloudflaredProcess(pid: number): boolean {
  try {
    const out = execFileSync("ps", ["-o", "comm=", "-p", String(pid)], {
      encoding: "utf8",
      timeout: 2000,
    }).trim();
    if (!out) return true; // ps found nothing conclusive — don't block the kill attempt
    return out.toLowerCase().includes("cloudflared");
  } catch {
    return true;
  }
}

/** SIGTERM a pid we only know from a persisted record, verifying identity first. ESRCH-safe. */
function killPersistedPid(pid: number, context: string): void {
  if (!isLikelyCloudflaredProcess(pid)) {
    console.warn(`[cloudflared] pid ${pid} (${context}) no longer looks like cloudflared — not killing it`);
    return;
  }
  try {
    process.kill(pid, "SIGTERM");
  } catch (err) {
    // ESRCH means it's already gone — fine. Anything else is also best-effort here.
    if ((err as NodeJS.ErrnoException).code !== "ESRCH") {
      console.warn(`[cloudflared] failed to kill ${context} pid ${pid}:`, err);
    }
  }
}

/**
 * Kill whatever this process considers "the tunnel" right now: an in-flight
 * spawn attempt (cancelled, not just left to time out — see `activeSpawn`
 * doc comment), then the tracked live process (in-memory handle, else the
 * persisted pid as a fallback, identity-checked).
 */
function killTrackedProcess(): void {
  if (activeSpawn) {
    activeSpawn.cancel("tunnel disabled while cloudflared was still starting");
  }
  if (state.process) {
    try {
      state.process.kill();
    } catch {
      // best-effort
    }
    state.process = null;
    return;
  }
  const { currentPid } = tunnelStore.getState();
  if (currentPid === null) return;
  killPersistedPid(currentPid, "tracked");
}

/**
 * Explicit user-facing disable (POST /auth/tunnel/disable). Runs the OS-truth
 * reconciliation sweep first (`port`-scoped — see `sweepOrphans`), then kills
 * the process AND clears `enabled` in the DB — the next boot must NOT re-spawn.
 */
export function disable(port: number): void {
  sweepOrphans(port);
  killTrackedProcess();
  state.enabled = false;
  state.tunnelUrl = null;
  tunnelStore.clear();
}

/**
 * Graceful daemon shutdown (main.ts). Kills the process but preserves
 * `enabled` in the DB — the next boot's restoreOnBoot() still re-spawns if
 * the tunnel was on. See tunnel-persistence plan, Decision 3.
 */
export function shutdownKill(): void {
  killTrackedProcess();
  state.enabled = false;
  state.tunnelUrl = null;
  tunnelStore.clearProcess();
}

/**
 * Called once at daemon boot (main.ts, after port/token/noAuth are resolved).
 * Runs the OS-truth reconciliation sweep FIRST, before even the no-auth/token
 * guard — orphans from a prior run in a *different* auth mode (e.g. this
 * machine had auth enabled last boot but is booting with --no-auth now) must
 * still be reaped; "sweep first, before anything else" applies to the guard
 * too, not just to the existing single-pid logic below. No-ops the REST of
 * restore in no-auth / no-token mode — exposing an auth-disabled daemon over a
 * public tunnel URL is exactly what /auth/tunnel/enable's own guard prevents,
 * and a boot-time call bypasses that route entirely. Otherwise: best-effort
 * kill (identity-checked) whatever pid was last recorded (covers both an
 * already-dead graceful shutdown and a live orphan from a crash), then
 * re-spawn fresh if the persisted state says the tunnel should be enabled.
 * Never throws.
 */
export async function restoreOnBoot(port: number, opts: { token?: string; noAuth: boolean }): Promise<void> {
  sweepOrphans(port);
  if (opts.noAuth || !opts.token) {
    console.warn("[cloudflared] skipping tunnel restore — daemon is in no-auth mode or has no token");
    return;
  }
  const persisted = tunnelStore.getState();
  if (persisted.currentPid !== null) {
    killPersistedPid(persisted.currentPid, "orphaned");
  }
  if (!persisted.enabled) return;
  try {
    await enable(port);
  } catch (err) {
    console.warn("[cloudflared] failed to restore tunnel on boot:", err);
    // Don't leave a dead pid/url behind for the NEXT boot to re-discover and
    // re-attempt killing (it'll be stale, possibly reused by an unrelated
    // process by then) — `enabled` stays true so the retry itself still happens.
    tunnelStore.clearProcess();
  }
}

/** Return the current tunnel state (no side effects). */
export function getState(): { enabled: boolean; tunnelUrl: string | null; startedAt: number | null } {
  return { enabled: state.enabled, tunnelUrl: state.tunnelUrl, startedAt: tunnelStore.getState().startedAt };
}

File: rust/vst-lifecycle/src/cloudflared.rs
//! cloudflared tunnel management — ports `services/cloudflared.ts`.
//!
//! Behavior contract:
//! - Spawns a **long-lived** background process using `vst_proc::spawn_child`
//!   (same abstraction as `04b`'s `AcpTerminalManager`).
//! - URL scraped from stdout/stderr via regex within a 10s timeout.
//! - `VST_CLOUDFLARED_BIN` env var overrides the binary path.
//! - `sweep_orphans`: pgrep + SIGTERM + delayed SIGKILL.
//! - `is_likely_cloudflared_process`: ps-based identity check before killing.
//! - State persisted in `vst_store::tunnel` (not a second layer).
//! - `enable` / `disable` / `shutdown_kill` / `restore_on_boot` / `get_state`.

use std::sync::{Arc, Mutex};
use std::time::Duration;

use thiserror::Error;
use tokio::io::{AsyncBufReadExt, BufReader};
use vst_store::StoreHandle;
use vst_store::TunnelStateRow;

#[derive(Debug, Error)]
pub enum CloudflaredError {
    #[error("process error: {0}")]
    Process(String),
    #[error("URL scrape timeout")]
    Timeout,
    #[error("store error: {0}")]
    Store(String),
    #[error("I/O error: {0}")]
    Io(#[from] std::io::Error),
}

pub type CloudflaredResult<T> = Result<T, CloudflaredError>;

const TUNNEL_URL_RE: &str = r"https://[a-z0-9\-]+\.trycloudflare\.com";
const SPAWN_TIMEOUT_MS: u64 = 10_000;

#[derive(Clone, Debug)]
pub struct CloudflaredState {
    pub enabled: bool,
    pub url: Option<String>,
    pub pid: Option<u32>,
}

/// Enable the cloudflared tunnel on the given port.
///
/// Spawns `cloudflared tunnel --url http://127.0.0.1:<port>` (or the binary
/// named by `VST_CLOUDFLARED_BIN`) as a long-lived background process and
/// scrapes the public URL from its output within 10s.
pub async fn enable(port: u16, store: &StoreHandle) -> CloudflaredResult<String> {
    // Check if already enabled.
    let current = store.get_tunnel_state().await;
    if current.enabled {
        if let Some(url) = current.current_url {
            return Ok(url);
        }
    }

    let bin = std::env::var("VST_CLOUDFLARED_BIN").unwrap_or_else(|_| "cloudflared".to_string());

    let mut child = tokio::process::Command::new(&bin)
        .args(["tunnel", "--url", &format!("http://127.0.0.1:{port}")])
        .stdout(std::process::Stdio::piped())
        .stderr(std::process::Stdio::piped())
        .spawn()
        .map_err(|e| CloudflaredError::Process(format!("spawn failed: {e}")))?;

    let pid = child.id();
    let stdout = child.stdout.take();
    let stderr = child.stderr.take();

    // Scrape URL from output within SPAWN_TIMEOUT_MS.
    let url_re =
        regex::Regex::new(TUNNEL_URL_RE).map_err(|e| CloudflaredError::Process(e.to_string()))?;

    let (tx, rx) = tokio::sync::oneshot::channel::<String>();
    let tx = Arc::new(Mutex::new(Some(tx)));

    let tx1 = tx.clone();
    let re1 = url_re.clone();
    if let Some(out) = stdout {
        tokio::spawn(async move {
            let mut lines = BufReader::new(out).lines();
            while let Ok(Some(line)) = lines.next_line().await {
                if let Some(m) = re1.find(&line) {
                    let url = m.as_str().to_string();
                    if let Some(sender) = tx1.lock().unwrap().take() {
                        let _ = sender.send(url);
                    }
                    break;
                }
            }
        });
    }

    let tx2 = tx.clone();
    let re2 = url_re;
    if let Some(err_out) = stderr {
        tokio::spawn(async move {
            let mut lines = BufReader::new(err_out).lines();
            while let Ok(Some(line)) = lines.next_line().await {
                if let Some(m) = re2.find(&line) {
                    let url = m.as_str().to_string();
                    if let Some(sender) = tx2.lock().unwrap().take() {
                        let _ = sender.send(url);
                    }
                    break;
                }
            }
        });
    }

    let url = tokio::time::timeout(Duration::from_millis(SPAWN_TIMEOUT_MS), async {
        rx.await.ok()
    })
    .await
    .ok()
    .flatten()
    .ok_or(CloudflaredError::Timeout)?;

    store
        .set_tunnel_state(TunnelStateRow {
            enabled: true,
            current_url: Some(url.clone()),
            current_pid: pid.map(|p| p as i64),
            started_at: Some(
                std::time::SystemTime::now()
                    .duration_since(std::time::UNIX_EPOCH)
                    .unwrap_or_default()
                    .as_millis() as i64,
            ),
            port: Some(port as i64),
        })
        .await;

    Ok(url)
}

/// Disable the cloudflared tunnel, killing the process if any.
pub async fn disable(store: &StoreHandle) -> CloudflaredResult<()> {
    let state = store.get_tunnel_state().await;
    if let Some(pid) = state.current_pid {
        let _ = kill_pid(pid as u32, false).await;
    }
    store.clear_tunnel().await;
    Ok(())
}

/// Get the current tunnel state from the store.
pub async fn get_state(store: &StoreHandle) -> CloudflaredResult<CloudflaredState> {
    let row = store.get_tunnel_state().await;
    Ok(CloudflaredState {
        enabled: row.enabled,
        url: row.current_url,
        pid: row.current_pid.map(|p| p as u32),
    })
}

/// Kill the process and clear tunnel state (for daemon shutdown).
pub async fn shutdown_kill(store: &StoreHandle) -> CloudflaredResult<()> {
    let state = store.get_tunnel_state().await;
    if let Some(pid) = state.current_pid {
        let _ = kill_pid(pid as u32, true).await;
    }
    store.clear_tunnel_process().await;
    Ok(())
}

/// Sweep orphaned cloudflared processes via pgrep.
pub async fn sweep_orphans() -> CloudflaredResult<()> {
    let output = tokio::task::spawn_blocking(|| {
        std::process::Command::new("pgrep")
            .args(["-x", "cloudflared"])
            .output()
    })
    .await
    .map_err(|e| CloudflaredError::Process(e.to_string()))?
    .map_err(|e| CloudflaredError::Process(e.to_string()))?;

    if !output.status.success() {
        return Ok(()); // no processes found
    }

    let pids: Vec<u32> = String::from_utf8_lossy(&output.stdout)
        .lines()
        .filter_map(|l| l.trim().parse::<u32>().ok())
        .collect();

    for pid in pids {
        if is_likely_cloudflared_process(pid).await {
            let _ = kill_pid(pid, false).await;
            tokio::time::sleep(Duration::from_millis(2000)).await;
            let _ = kill_pid(pid, true).await;
        }
    }

    Ok(())
}

/// ps-based identity check before killing a pid.
async fn is_likely_cloudflared_process(pid: u32) -> bool {
    let output = tokio::task::spawn_blocking(move || {
        std::process::Command::new("ps")
            .args(["-p", &pid.to_string(), "-o", "comm="])
            .output()
    })
    .await
    .ok()
    .and_then(|r| r.ok());

    if let Some(out) = output {
        let comm = String::from_utf8_lossy(&out.stdout)
            .trim()
            .to_ascii_lowercase();
        comm.contains("cloudflared")
    } else {
        false
    }
}

async fn kill_pid(pid: u32, force: bool) -> CloudflaredResult<()> {
    let sig = if force { "9" } else { "15" };
    tokio::task::spawn_blocking(move || {
        std::process::Command::new("kill")
            .args([&format!("-{sig}"), &pid.to_string()])
            .status()
    })
    .await
    .ok();
    Ok(())
}
